Document rpi-connect 2.1.0's new doctor CLI (#3910)
raspberrypi/documentation: documentation/asciidoc/services/connect/troubleshooting.adoc

==== Networking and firewall issues

> Connect avoids changing your network or firewall in order to establish a connection. However, if you have a restrictive network and experience trouble connecting, `rpi-connect` and `rpi-connect-lite` include a `rpi-connect doctor` command to run a series of tests to check that Connect can establish connections properly.

Reviewer: ```suggestion
Connect usually communicates between devices without requiring changes to your network or firewall. However, especially restrictive networks can sometimes block Connect communication. To help debug problems with such networks, `rpi-connect` and `rpi-connect-lite` include the `rpi-connect doctor` command. `rpi-connect doctor` runs a series of tests to check that Connect communication functions properly on your network.
```

Specifically:

> . Can we make an HTTPS request to the Raspberry Pi Connect API at `api.connect.raspberrypi.com` on port 443 and have it respond successfully?

Reviewer 2: Is it worth clarifying that this first request is unauthenticated?

Reviewer: discussing below!

> Specifically:
> 
> . Can we make an HTTPS request to the Raspberry Pi Connect API at `api.connect.raspberrypi.com` on port 443 and have it respond successfully?
> . Can we make an authenticated request to the Raspberry Pi Connect API?
> . Can we communicate with the Raspberry Pi Connect STUN or TURN servers at `stun.raspberrypi.com`, `turn1.raspberrypi.com`, `turn2.raspberrypi.com`, or `turn3.raspberrypi.com` on UDP port 3478 and receive a server reflexive candidate?
> . Can we communicate with the Raspberry Pi Connect TURN servers on TCP ports 3478 or 443, and UDP ports 3478, 443, and 49152 -> 65535 and receive a relay candidate?

Reviewer: ```suggestion
```

> To run these tests on your device, run the following command:
> 
> [source,console]
> ----
> $ rpi-connect doctor
> ✓ Communication with Raspberry Pi Connect API
> ✓ Authentication with Raspberry Pi Connect API
> ✓ Peer-to-peer connection candidate via STUN
> ✓ Peer-to-peer connection candidate via TURN
> ----

Reviewer: ```suggestion
To run these tests on your device, run the following command:

[source,console]
----
$ rpi-connect doctor
----

If Connect can communicate properly on your network, you should see output similar to the following:

----
✓ Communication with Raspberry Pi Connect API
✓ Authentication with Raspberry Pi Connect API
✓ Peer-to-peer connection candidate via STUN
✓ Peer-to-peer connection candidate via TURN
----

If Connect can't communicate properly on your network, you'll see an "x" instead of a check next to the failing test case. Ask your network administrator to enable the following connections on your network:

* unauthenticated HTTPS requests to the Raspberry Pi Connect API on port 443 of `api.connect.raspberrypi.com` 
* authenticated requests to the Raspberry Pi Connect API on port 443 of `api.connect.raspberrypi.com`
* requests to Raspberry Pi Connect STUN or TURN servers on UDP port 3478 of at least one of the following:
** `stun.raspberrypi.com`
** `turn1.raspberrypi.com`
** `turn2.raspberrypi.com`
** `turn3.raspberrypi.com`
* requests to Raspberry Pi Connect TURN servers on TCP ports 3478 or 443 of at least one of the following:
** `turn1.raspberrypi.com`
** `turn2.raspberrypi.com`
** `turn3.raspberrypi.com`
* requests to Raspberry Pi Connect TURN servers on UDP ports 3478, 443, or 49152 -> 65535 of at least one of the following:
** `turn1.raspberrypi.com`
** `turn2.raspberrypi.com`
** `turn3.raspberrypi.com`
```

Reviewer 2: I believe that as far as firewalls / network-blocking goes, there's no difference between an authenticated or unauthenticated request to port 443? (but obviously @mudge will be able to confirm!)

Author: That's right, from a networking point of view, the authentication isn't relevant though from an application point of view, you're going to have trouble if you're signed in to Connect and the authenticated request fails.

Re "at least one of the following": it might be better to say "all of the following" as Connect will always use stun.raspberrypi.com and will use any one of turn1, turn2 and turn3 (depending on where you are and what the API is currently returning).

Reviewer: That was my suspicion as well, and part of why I decided to 'poke the bear' with this suggestion. We should either:

* remove the second bullet point because it's redundant and remove the 'unauthenticated' from the first bullet
* clarify that the second bullet is HTTP or HTTPS

Reviewer: Technically, for basic functionality, folks only need one, right? Or should we just say 'all' because it's silly to only open one?

Reviewer: Also, since I don't have any failing Connect test cases on hand: is the 'x' explanation correct? Should users expect to see any other text there? If you have an example fail case, I could add it here.

Reviewer: @mudge can you clarify what we should do for the first two bullet points? Should I merge them into a single point, or do you have a preferred wording to distinguish them better?

Author: If your goal is to advise what traffic to allow to api.connect.raspberrypi.com, I'd merge them into a single point and not mention authentication. rpi-connect doctor will perform two separate tests but the networking-level fix is just one point.
